esp32/modmachine: fix reset_cause

tve opened the pull request:
Fixes #5134
The code used to call `rtc_get_reset_reason` which is a "raw" reset cause. ESP-IDF massages that for the proper reset cause available from `esp_reset_cause`.
I would like to add a `machine.PANIC_RESET` to be able to distinguish a clean reset for seomthing like OTA update from an abort/assert type of reset, dunno whether that's acceptable.

tve committed:
esp32/modmachine: fix reset_cause

tve committed:
rejiggled two reset causes

tve commented:
I plan to add a unit test...

tve committed:
codeformat of modmachine

dpgeorge commented:
Sorry this got lost... it looks good to me, I can merge it if it's ready?

tve commented:
It's ready as in: it works. I wanted to write a unit test. Also, there's the open question whether it would be good to add a `machine.PANIC_RESET` to capture the difference between SW calling reset vs a panic-crash causing reset.
('been busy with other projects this past week)

dpgeorge commented:
Looking at this again, the SOFT_RESET cause is not correct: that code is supposed to be for a ctrl-D reset at the REPL.  The codes are:
```
PWRON_RESET - device powered up
HARD_RESET - device fully reset after already being powered up, eg via machine.reset()
WDT_RESET - reset due to machine.WDT() (or other WDT) resetting the device
DEEPSLEEP_RESET - wake from machine.deepsleep()
SOFT_RESET - MicroPython ctrl-D reset at the REPL
```

> Also, there's the open question whether it would be good to add a `machine.PANIC_RESET` to capture the difference between SW calling reset vs a panic-crash causing reset.

Adding new codes means all ports should add and implement them... but if it's useful it could be added.  Maybe `PANIC_RESET` is an unintended `HARD_RESET`?

tve commented:
I'll fix the SOFT_RESET.

The question about PANIC_RESET is whether it's worth distinguishing between an intentional call to machine.reset() for example to update the code being run vs. a reset caused by a panic type of abort. There are definitely use-cases where I can see it matter. On the other hand one can argue that the distinction can be made by leaving some value in some non-cleared memory location. If you want the reset causes to be the same in all ports, then maybe best not to rock that boat here.

dpgeorge commented:
> On the other hand one can argue that the distinction can be made by leaving some value in some non-cleared memory location

Yes, like the RTC memory, or in a file on the filesystem.

At this stage I'd suggest to not add new codes and just fix the existing behaviour as per this PR.  New codes can be discussed separately (maybe we can allow custom codes for ports...).

tve committed:
esp32/machine: ensure reset causes match other ports

tve commented:
I removed the incorrect SOFT_RESET. I also moved the Brownout into PWRON_RESET 'cause that's what it resembles the most as far as I know. I tested the cases manually 'cause a unit test is a bit of a mess and I don't think that's worth it for this function since it just dispatches on a value provided by esp-idf.

dpgeorge commented:
Thanks, looks good now, squashed and merged in 50a7ba234850183693d2b236e48ec0b76399123d
